src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        protected virtual Task<TResponse> RemapResponseAsync(TResponse delegatedResponse, RazorCodeDocument codeDocument, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// Extracts information about the request and then calls <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>.
+        /// Children can override to provide any more custom logic needed before doing a normal delegated response pass with <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
+
+            return await GetDelegatedResponseAsync(request, documentContext, codeDocument, sourceText, absoluteIndex, projection, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected async Task<TResponse?> GetDelegatedResponseAsync(
+            TRequest request,
+            DocumentContext documentContext,
+            RazorCodeDocument codeDocument,
+            SourceText sourceText,
+            int absoluteIndex,
+            Projection projection,
+            CancellationToken cancellationToken)
+        {
+            if (!_languageServerFeatureOptions.SingleServerSupport)
+            {
+                return await HandleWithoutSingleServerAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (projection.LanguageKind == RazorLanguageKind.Razor)
+            {
+                return await HandleInRazorAsync(request, documentContext, codeDocument, sourceText, cancellationToken).ConfigureAwait(false);
```

Comment:
This feels wrong... why are we handling razor in a method called "Get*Delegated*Response"? 

I would expect this call to be in the Handle method, and then a check for single server, and if single server is on, _then_ this method is called to delegate.

Reply:
This may be where the fallback behavior is fuzzy to me. Is handling in razor the same as not delegating at all to c#/html?

Reply:
In a sense. Essentially this endpoint is responsible for providing hover information about Razor tags. If single server is on it _also_ delegates to C#/HTML servers if there is nothing it can provide. If single server is not on, then it doesn't need to delegate (as the other server will respond), but it still needs to provide razor info.

Reply:
I removed this method entirely, see what you think now